### PR TITLE
Closed captioning support

### PR DESCRIPTION
**Pull request by takecare**
This PR adds the capability to check if a device supports closed captioning. A compat-style class is used to assure we're able to check for the closed captioning in devices running lower that KIT KAT (19).

### CODE REVIEW

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    public static CaptionManager newInstance(Context context) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
+            return ClosedCaptionManager.newInstance(context);
+        } else {
+            return DummyClosedCaptionManager.newInstance();
+        }
+    }
+
+    private static class ClosedCaptionManager implements CaptionManager {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java]: The class should be marked as final too if the ctor is private, same below

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package com.novoda.accessibility;
+
+import android.annotation.TargetApi;
+import android.content.Context;
+import android.os.Build;
+import android.view.accessibility.CaptioningManager;
+
+public class ClosedCaptionManagerCompat {
+
+    public static CaptionManager newInstance(Context context) {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java]: :x: if ClosedCaptionManagerCompat doesn't implement CaptionManager, then this should not be called `newInstance(Context)`; I would expect static factory methods called `newInstance` to return an instance of that class.

It should be something like `createCaptionManager(Context)` or `newCaptionManager(Context)`

File: core/src/main/java/com/novoda/accessibility/AccessibilityServices.java [hunk 3]
```diff
@@ -33,4 +38,10 @@ public boolean isSpokenFeedbackEnabled() {
         return accessibilityManager.getEnabledAccessibilityServiceList(feedbackTypeFlags);
     }
 
+    /**
+     * Reports if video captioning is enabled on the device.
+     */
+    public boolean isClosedCaptioningEnabled() {
+        return captionManager.isClosedCaptioningEnabled();
+    }
```
Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/AccessibilityServices.java:46]: needs a space after this

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
+            return ClosedCaptionManager.newInstance(context);
+        } else {
+            return DummyClosedCaptionManager.newInstance();
+        }
```
Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerCompat.java]: needs a space after this (same with the next class)

**Comment by ataulm** [comment]
:100: nice one man, only blocking thing is `newInstance` method name -> would rename to something that highlights it's not creating a new instance of ClosedCaptioningManagerCompat.

**Comment by ataulm** [comment]
:+1:

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package com.novoda.accessibility;
+
+import android.annotation.TargetApi;
+import android.content.Context;
+import android.os.Build;
+import android.view.accessibility.CaptioningManager;
+
+public class ClosedCaptionManagerFactory {
+
+    public static CaptionManager createCaptionManager(Context context) {
```
Comment (rock3r) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:10]: What about `from`?

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
```
Comment (rock3r) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java]: What about `osSupportsNativeClosedCaptioning()`?

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        } else {
+            return DummyClosedCaptionManager.newInstance();
+        }
+    }
+
+    private static final class ClosedCaptionManager implements CaptionManager {
+
+        private final CaptioningManager captioningManager;
+
+        @TargetApi(Build.VERSION_CODES.KITKAT)
```
Comment (rock3r) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:26]: Could be applied to the whole class, probably. Even though, narrower scope and all.

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        @Override
+        public boolean isClosedCaptioningEnabled() {
+            return false;
+        }
+
+    }
+
+    interface CaptionManager {
```
Comment (rock3r) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:61]: Either this interface should be public, or the factory should be package private :dancer: I suppose the latter.

**Comment by rock3r** [comment]
Few comments, then LGTM.

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java (hunk 8 above)
Comment (takecare) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:61]: latter :+1:

File: core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java (hunk 5 above)
Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:10]: :rage4: noooooooooooo

Comment (ataulm) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:10]: to the from. `from` _only only only_ makes sense if **all** of the following hold:
- it's a static method
- the class that the static method is in is the type that you're converting _to_
- the single parameter you pass to the static method is the type that you're converting _from_

Comment (takecare) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:10]: `createCaptionManagerFrom` or leave it then?

Comment (rock3r) [inline core/src/main/java/com/novoda/accessibility/ClosedCaptionManagerFactory.java:10]: Or `with` as some libraries seem to be doing, but that's usually for builders, so dunno.

**Comment by takecare** [comment]
@ataulm @rock3r comments addressed

**Pull request merged**